arviz/kde_utils.py

```diff
@@ -0,0 +1,863 @@
+
+    References
+    ----------
+    .. [1] Kernel density estimation via diffusion.
+       Z. I. Botev, J. F. Grotowski, and D. P. Kroese.
+       Ann. Statist. 38 (2010), no. 5, 2916--2957.
+    """
+
+    k_sq = np.asfarray(k_sq, dtype=np.float64)
+    a_sq = np.asfarray(a_sq, dtype=np.float64)
+
+    l = 7
+    f = np.sum(np.power(k_sq, l) * a_sq * np.exp(-k_sq * np.pi ** 2 * t))
+    f *= 0.5 * np.pi ** (2.0 * l)
+
+    for j in reversed(range(2, l)):
+        c1 = (1 + 0.5 ** (j + 0.5)) / 3
+        c2 = np.product(np.arange(1.0, 2 * j + 1, 2, dtype=np.float64))
+        c2 /= (np.pi / 2) ** 0.5
+        t_j = np.power((c1 * (c2 / (N * f))), (2.0 / (3.0 + 2.0 * j)))
+        f = np.sum(k_sq ** j * a_sq * np.exp(-k_sq * np.pi ** 2.0 * t_j))
+        f *= 0.5 * np.pi ** (2 * j)
+
+    out = t - (2 * N * np.pi ** 0.5 * f) ** (-0.4)
+    return out
+
+
+def _root(function, N, args, x):
+    # The right bound is at most 0.01
+    found = False
+    N = max(min(1050, N), 50)
+    tol = 10e-12 + 0.01 * (N - 50) / 1000
+
+    while not found:
+        try:
+            bw, res = brentq(function, 0, 0.01, args=args, full_output=True, disp=False)
+            found = True if res.converged else False
+        except ValueError:
+            bw = 0
+            tol *= 2.0
+            found = False
+        if bw <= 0 or tol >= 1:
+            # warnings.warn(
+            #    "Improved Sheather-Jones did not converge as expected. "
+            #    "Using Silverman's rule instead.",
+            #    Warning
+            # )
+            bw = (_bw_silverman(x) / np.ptp(x)) ** 2
+            return bw
+    return bw
+
+
+def _check_type(x):
+    """
+    Checks the input is of the correct type.
+    It only accepts numeric lists/numpy arrays of 1 dimension or something that
+    can be flattened to 1 dimension.
+
+    Parameters
+    ----------
+    x : Object whose type is checked before computing the KDE.
+
+    Returns
+    -------
+    x : 1-D numpy array
+        If no error is thrown, a 1 dimensional array of
+        sample data from the variable for which a density estimate is desired.
+
+    """
+    if not isinstance(x, (list, np.ndarray)):
+        raise ValueError((
+            f"Can't produce a density estimator for {type(x)}.\n"
+            f"Please input a list with numbers or numpy array."
+        ))
+    # Will raise an error if `x` can't be casted to numeric or flattened to one dimension.
+    try:
+        x = np.asfarray(x).flatten()
+    except Exception as e:
+        print(
+            "The following exception occurred while trying to convert `x`",
+            "to a 1 dimensional float array."
+        )
+        raise e
+
+    x = x[np.isfinite(x)]
+
+    if x.size == 0:
+        raise ValueError("`x` does not contain any finite number.")
+    if x.size == 1:
+        raise ValueError("`x` is of length 1. Can't produce a KDE with only one data point.")
```

Comment:
should 2 data points be allowed? maybe we should set a threshold somewhere and between 2 and threshold issue a warning?

Reply:
I think it should be allowed. Both because this was the previous behavior and because it's the minimum number for the method to work. We could have threshold in other places, like we already have in plot_ppc or as we could add to plot_dist_comparison

Reply:
An initial version threw a warning when the size was below 30 or 50 but then I decided to remove it. I think I did it because of the examples using eight schools data (which would have caused warnings everywhere). I don't have any problem in adding warnings, but first I would like to check if others have any opinions on this.